microsoft-edge/devtools-guide-chromium/memory-problems/heap-snapshot-schema.md

```diff
@@ -0,0 +1,314 @@
+| Native | Things allocated by the Blink rendering engine, not by V8. These are mostly DOM: HTMLDivElement, CSSStyleRule, etc. |
+| Concatenated string | The result of concatenating two strings with the `+` operator. Rather than making a new string with a copy of all the data from the two source strings, V8 creates a ConsString object with pointers to the two source strings. From a JavaScript perspective, it acts just like any other string, but from a memory profiling perspective, it is different. |
+| Sliced string | The result of a substring operation, such as using `String.prototype.substr` or `String.prototype.substring`. V8 avoids copying string data by instead creating a SlicedString which points to the original string and specifies the start index and length. From a JS perspective, it acts just like any other string, but from a memory profiling perspective, it is different. |
+| Array | Various internal lists, which show up in the devtools with the category name **(array)**. Like Hidden, this category clumps together a bunch of unrelated things. Many of the objects here are named **(object properties)** or **(object elements)**, indicating that they contain the string-keyed or numeric-keyed properties of a JavaScript object. |
+| Code | Things that grow proportionally to the amount of script, and/or the number of times that functions run. |
+| Synthetic | Synthetic nodes don't correspond to anything actually allocated in memory. These are used to distinguish the different kinds of GC roots. |
+
+### Edges
+
+Similar to the `nodes` array, the `edges` top-level array contains all the individual elements needed to recreate the edges of the graph.
+
+Also similar to nodes, the total number of edges can be calculated by multiplying `snapshot.edge_count` with `snapshot.meta.edge_fields.length`. Edges are also stored as a sequence of numbers which you will need to iterate on by groups of `snapshot.meta.edge_fields.length`.
+
+However, to read the `edges` array correctly, you first need read the `nodes` array because each node knows how many edges it has. 
+
+To recreate an edge you need three pieces of information:
+
+* The edge type.
+* The edge name or index.
+* The node the edge is connected to.
+
+You can use `snapshot.meta.edge_types` to understand what each element is.
+
+For example, if you read the first node in the `nodes` array and its `edge_count` property is set to 4, then the first 4 groups of `snapshot.meta.edge_fields.length` numbers in the `edges` array correspond to the 4 edges of this node.
+
+| Index in edge group | Name | Description |
+| --- | --- | --- |
+| `0` | `type` | The type of edge. See [Edge types](#edge-types) to find out what are the possible types. |
+| `1` | `name_or_index` | This can be a number or a string. If it’s a number, it corresponds to the index in the top-level `strings` array where the name of the edge can be found. |
```

Comment:
```suggestion
| `1` | `name_or_index` | This can be a number or a string. If it's a number, it corresponds to the index in the top-level `strings` array where the name of the edge can be found. |
```
tip: if Doc Writer Pack extension-set is installed in VS Code, to auto straighten all quotes:
Ctrl+A, Cut
Paste
Save
See delta in a diff view eg GitHub Desktop, or VS Code's Source Control view.